### ISSUE TITLE
explore 'recursion' crate, or at least copy its algorithms

### ISSUE THREAD

**Issue by apoelstra** [rust-bitcoin/rust-miniscript#484]
See this blog post series: https://recursion.wtf/posts/rust_schemes_3/ which describes a scheme for managing recursive data structures which is likely to be (much) faster than our existing code and also avoid stack overflows.

**Comment by sanket1729** [comment]
I will explore this weekend. At the very least, we should be able to copy the algorithms.

**Comment by apoelstra** [comment]
A couple other related ideas bumping around in my brain:

* @JeremyRubin has suggested we have a `MiniscriptView`/`PolicyView` etc type which is a tree that we users can use to pattern-match with etc, but which can't be directly constructed (so that you can't construct Miniscript programs that don't typecheck)
* Relatedly [this HN comment](https://news.ycombinator.com/item?id=24226048) describes a way using arenas that we can use `&`-pointers rather than `Box`-pointers which makes matching much much easier.

So on a high level here's what I'm imagining ... basically `Policy` becomes backed by a `Vec` of `PolicyNode`s (the blog post uses "Layer" for what we call "Node"); each `PolicyNode` is an enum where the variants contain `&`-refs to other `PolicyNodes`. These `&`-refs will point into the same vector; we will use `Pin` somehow to achieve this. Then we have a `.view(&self) -> &PolicyNode` method which returns a ref to the topmost `PolicyNode` which the user can easily match on.

Obviously the devil is in the details :)

**Comment by apoelstra** [comment]
I think the `Layer` terminology comes from https://blog.sumtypeofway.com/posts/introduction-to-recursion-schemes.html which the `recursion` crate is based on, and is probably worth reading in its entirety (even though it's a 5-part series, written in Haskell, and uses Haskell features that don't nicely map to Rust)

**Comment by apoelstra** [comment]
cc @sipa in case you are interested in any of this (if not feel free to unsubscribe :P)

**Comment by apoelstra** [comment]
Sounds good, @sipa. I think if we continue to use naive recursive data structures this is the way we'll wind up going. Though there are a few limitations -- one is about efficiency, since it's slow to make a zillion tiny allocations/deallocations and also bad for cache locality -- but another, more Rust-specific, is about the ergonomics of matching on recursive structures.

Then there is the issue that we'd like to expose library code to allow construction and pattern-matching on arbitrary miniscripts without allowing people to construct things that don't typecheck (since these will eventually trigger assertions if you try to poke them too much).

**Comment by apoelstra** [comment]
Adding one more link to this mess of an issue: [this reddit post about generalizing over Rc and Arc](https://www.reddit.com/r/rust/comments/6dz0xh/abstracting_over_reference_counted_types_rc_and/) has an interesting "double-trait" trick which lets you get something like an infinite type which may allow us to create a fixpoint type analogous to the one in the Haskell blog post.

**Comment by apoelstra** [comment]
After spending a few hours on this, I think

* Abstracting over recursion is cool, but the proposal in the linked blog post still depends on using boxes for recursive structures, which is (a) inefficient and (b) won't let us do things like "view types" that can be easily matched on. What it **will** do is take a ton of engineering time and make it really hard to later change the backing store to be something flat...
* ...so, I think we want to do backing store efficiency improvements first
* Essentially, what this would involve is allocating a slab for nodes and then storing a raw pointer to this inside of the API-exposed `Miniscript` structure (and `Policy` etc would work the same). Annoyingly we have to use a raw pointer ... using a `Vec<T>`, `Box<[T]>` or even a `*mut [T]` all run into limitations related to the "stacked borrows" aliasing rules in Rust, either because things are forbidden or they're simply undefined-as-of-now.
* Adding insult to injury, it's hard to do a lot of stuff with raw pointers into slabs with 1.41. In particular the `offset_from` method on pointers (basically, pointer subtraction) wasn't added til Rust 1.46. So if you want to mess around with pointers -- e.g. to map over all nodes in a tree and reconstituting the pointers in the map output -- you have to cast to integers, do math there, and cast back ... which again runs into poorly-defined areas of stacked borrows.

Basically let's just revisit this after the next MSRV bump, and if we're lucky we'll have a better-defined language by then. Then we should start by defining a generic slab-backed tree type, use these Haskell blog posts as inspiration for what APIs we want to define on that, and then try porting our various trees into it. It wouldn't surprise me if we can eventually 10x the performance of this library while making it easier to add functionality.

As an aside, one way I considered to bypass Rust's memory model is to drop into C :}. But because we're trying to construct complicated Rust enums it would be very difficult to pass these things across the FFI barrier, not to mention the heavy optimization cost we'd face.

**Comment by apoelstra** [comment]
Relevant (about ergonomically bundling a slab with a recursive type that points into it) https://old.reddit.com/r/rust/comments/zc8pq0/selfreferential_types_for_fun_and_profit/izp774p/

**Comment by apoelstra** [comment]
This one is also highly relevant to rust-miniscript and its context parameter (and most of my attempts to de-recursive-ify the data structures):

http://ais523.me.uk/blog/scoped-generics.html
